### ISSUE TITLE
error starting Dynamodb on OS X 10.12.6

### ISSUE THREAD

**Issue by noah10** [localstack/localstack#476]
While I've seen variants on this error reported, the suggested solutions that might be applicable (e.g. un-installing and re-installing localstack) haven't fixed this, so I'm opening a new issue. Here's what I get when I try to run dynamodb with localstack:

```
SERVICES=dynamodb DATA_DIR=/tmp/localstack/data DEBUG=1 DEFAULT_REGION=us-east-2 localstack start
Starting local dev environment. CTRL-C to quit.
Starting mock DynamoDB (http port 4569)...
Initializing DynamoDB Local with the following configuration:
Port:	4564
InMemory:	false
DbPath:	/tmp/localstack/data/dynamodb
SharedDb:	true
shouldDelayTransientStatuses:	false
CorsParams:	*

2017-11-21T11:22:20:ERROR:localstack.services.generic_proxy: Error forwarding request: HTTPConnectionPool(host='127.0.0.1', port=4564): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x111034210>: Failed to establish a new connection: [Errno 61] Connection refused',)) Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/localstack/services/generic_proxy.py", line 181, in forward
    headers=forward_headers)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 112, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/adapters.py", line 508, in send
    raise ConnectionError(e, request=request)
ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=4564): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x111034210>: Failed to establish a new connection: [Errno 61] Connection refused',))

Ready.

```

The presence or absence of DATA_DIR and DEFAULT_REGION makes no difference. FWIW, I do *not* get that error if I run it with --docker.

Here's some other info that might be helpful in diagnosing the problem:

```
$ localstack --version
0.8.2
$ node -v
v8.6.0
$ npm -v
5.5.1
$ ls -la /usr/local/lib/python2.7/site-packages/localstack/infra/dynamodb/
total 7288
drwxr-xr-x   7 noah  admin      238 Nov 21 09:54 .
drwxr-xr-x   6 noah  admin      204 Nov 21 09:55 ..
-rw-r--r--   1 noah  admin  3713946 Nov 21 09:54 DynamoDBLocal.jar
drwxr-xr-x  51 noah  admin     1734 Nov 21 09:54 DynamoDBLocal_lib
-rw-r--r--   1 noah  admin     8644 Nov 21 09:54 LICENSE.txt
-rw-r--r--   1 noah  admin      795 Nov 21 09:54 README.txt
drwxr-xr-x  25 noah  admin      850 Nov 21 09:54 third_party_licenses
$ java -jar /usr/local/lib/python2.7/site-packages/localstack/infra/dynamodb/DynamoDBLocal.jar -port 4564
Initializing DynamoDB Local with the following configuration:
Port:	4564
InMemory:	false
DbPath:	null
SharedDb:	false
shouldDelayTransientStatuses:	false
CorsParams:	*
```
(Just to be clear, running the DynamoDB jar by itself doesn't seem to produce any errors.)

**Comment by xuru** [comment]
I can confirm this as well.
```
System Version:	macOS 10.12.6 (16G1036)
Kernel Version:	Darwin 16.7.0
```
```
$ DEBUG=1 localstack start
Starting local dev environment. CTRL-C to quit.
Starting mock SQS (http port 4576)...
Starting mock DynamoDB (http port 4569)...
Initializing DynamoDB Local with the following configuration:
Port:	4564
InMemory:	false
DbPath:	/tmp/localstack/data/dynamodb
SharedDb:	true
shouldDelayTransientStatuses:	false
CorsParams:	*

15:49:35.622 [main] INFO  org.elasticmq.server.Main$ - Starting ElasticMQ server (0.13.8) ...
15:49:36.191 [elasticmq-akka.actor.default-dispatcher-3] INFO  akka.event.slf4j.Slf4jLogger - Slf4jLogger started
15:49:36.928 [elasticmq-akka.actor.default-dispatcher-3] INFO  o.e.rest.sqs.TheSQSRestServerBuilder - Started SQS rest server, bind address 0.0.0.0:4561, visible server address http://xuru.mbp.local:4576
15:49:36.929 [main] INFO  org.elasticmq.server.Main$ - === ElasticMQ server (0.13.8) started in 1795 ms ===
2017-11-22T15:50:53:WARNING:infra.py: Service "dynamodb" not yet available, retrying...
2017-11-22T15:52:12:WARNING:infra.py: Service "dynamodb" not yet available, retrying...
2017-11-22T15:53:30:WARNING:infra.py: Service "dynamodb" not yet available, retrying...
2017-11-22T15:54:49:WARNING:infra.py: Service "dynamodb" not yet available, retrying...
2017-11-22T15:56:08:WARNING:infra.py: Service "dynamodb" not yet available, retrying...
```
```
$ localstack --version
0.8.2
$ java -version
java version "1.8.0_112"
Java(TM) SE Runtime Environment (build 1.8.0_112-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.112-b16, mixed mode)
```
I made sure to ensure I had the latest java version.

**Comment by xuru** [comment]
Looks like I have the same issue using docker:
```
$ DEBUG=1 localstack start --docker
Starting local dev environment. CTRL-C to quit.
docker run -it -e SERVICES="dynamodb,sqs" -e HOSTNAME="xuru.mbp.local" -e LOCALSTACK_HOSTNAME="xuru.mbp.local" -e DEBUG="1" -e DATA_DIR="/tmp/localstack_data" -p 8080:8080  -p 4569:4569 -p 4576:4576 -v "/tmp/localstack/data:/tmp/localstack_data" -v "/private/var/folders/dp/k1d4s3ns6djcc_psfg__6h440000gn/T/localstack:/tmp/localstack" -v "/var/run/docker.sock:/var/run/docker.sock" -e DOCKER_HOST="unix:///var/run/docker.sock" -e HOST_TMP_FOLDER="/private/var/folders/dp/k1d4s3ns6djcc_psfg__6h440000gn/T/localstack" "localstack/localstack"
2017-11-22 22:13:05,008 CRIT Supervisor running as root (no user in config file)
2017-11-22 22:13:05,011 INFO supervisord started with pid 1
2017-11-22 22:13:06,014 INFO spawned: 'dashboard' with pid 9
2017-11-22 22:13:06,017 INFO spawned: 'infra' with pid 10
(. .venv/bin/activate; bin/localstack web)
(. .venv/bin/activate; exec bin/localstack start)
Starting local dev environment. CTRL-C to quit.
2017-11-22T22:13:06:INFO:werkzeug:  * Running on http://0.0.0.0:8080/ (Press CTRL+C to quit)
2017-11-22T22:13:06:INFO:werkzeug:  * Restarting with stat
2017-11-22T22:13:07:WARNING:werkzeug:  * Debugger is active!
2017-11-22 22:13:07,246 INFO success: dashboard entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2017-11-22 22:13:07,246 INFO success: infra entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2017-11-22T22:13:07:INFO:werkzeug:  * Debugger PIN: 966-726-561
Starting mock DynamoDB (http port 4569)...
Starting mock SQS (http port 4576)...
Initializing DynamoDB Local with the following configuration:
Port:	4564
InMemory:	false
DbPath:	/tmp/localstack_data/dynamodb
SharedDb:	true
shouldDelayTransientStatuses:	false
CorsParams:	*

22:13:09.260 [main] INFO  org.elasticmq.server.Main$ - Starting ElasticMQ server (0.13.8) ...
22:13:10.114 [elasticmq-akka.actor.default-dispatcher-2] INFO  akka.event.slf4j.Slf4jLogger - Slf4jLogger started
22:13:11.382 [elasticmq-akka.actor.default-dispatcher-3] INFO  o.e.rest.sqs.TheSQSRestServerBuilder - Started SQS rest server, bind address 0.0.0.0:4561, visible server address http://xuru.mbp.local:4576
22:13:11.382 [main] INFO  org.elasticmq.server.Main$ - === ElasticMQ server (0.13.8) started in 2984 ms ===
2017-11-22T22:14:26:WARNING:infra.pyc: Service "dynamodb" not yet available, retrying...
2017-11-22T22:15:45:WARNING:infra.pyc: Service "dynamodb" not yet available, retrying...
```
```
$ docker info
Containers: 31
 Running: 2
 Paused: 0
 Stopped: 29
Images: 227
Server Version: 17.11.0-ce-rc4
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 992280e8e265f491f7a624ab82f3e238be086e49
runc version: 0351df1c5a66838d0c392b4ac4cf9450de844e2d
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.60-linuxkit-aufs
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.855GiB
Name: linuxkit-025000000001
ID: 77M3:MSHY:AJM5:WNCN:PNKP:FL34:PTGB:FBZ2:OBNG:QRAQ:F5ZX:7N2J
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 30
 Goroutines: 57
 System Time: 2017-11-22T22:14:01.646624678Z
 EventsListeners: 2
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```
```
Version 17.11.0-ce-rc4-mac39 (20423)
Channel: edge
e557194136
```
```
$ cat .env
#!/usr/bin/env bash
export SERVICES=dynamodb,sqs
export DATA_DIR=/tmp/localstack/data
export HOSTNAME=localhost
export HOSTNAME_EXTERNAL=localhost
```

**Comment by whummer** [comment]
Thanks for reporting @noah10 . The error output you are seeing is misleading. The stack actually starts up properly, as you can see from the output line `Ready.`. The (temporary) error is caused by a timing issue where we start up the services and immediately start health checking them by querying the service endpoints (see [here](https://github.com/localstack/localstack/blob/d797626ae64cc994200e1186dfa616f13c7599ca/localstack/services/dynamodb/dynamodb_starter.py#L17-L18)).

This has been fixed in #483 where we now wait until the network port becomes available before running the health check probe requests. You should not be seeing this error anymore during startup. I'm closing this issue, but please feel free to re-open or leave a comment here if the problem persists. Thanks
